staging/src/k8s.io/apiserver/pkg/endpoints/responsewriter/wrapper_test.go

```diff
@@ -0,0 +1,311 @@
+					err := recover()
+					switch {
+					case err != nil:
+						if !test.panicExpected {
+							t.Errorf("Expected no panic, but got: %v", err)
+						}
+					default:
+						if test.panicExpected {
+							t.Errorf("Expected a panic")
+						}
+					}
+				}()
+
+				originalGot := GetOriginal(wrapped)
+				if originalExpected != originalGot {
+					t.Errorf("Expected to get tehe original http.ResponseWriter object")
+				}
+			}()
+		})
+	}
+}
+
+func newServer(t *testing.T, h http.Handler, http2 bool) *httptest.Server {
+	server := httptest.NewUnstartedServer(h)
+	switch {
+	case http2:
+		server.EnableHTTP2 = true
+		server.StartTLS()
+	default:
+		server.Start()
+	}
+	_, err := url.Parse(server.URL)
+	if err != nil {
+		t.Fatalf("Expected the server to have a valid URL, but got: %s", server.URL)
+	}
+	return server
+}
+
+func sendRequest(t *testing.T, server *httptest.Server) {
+	req, err := http.NewRequest("GET", server.URL, nil)
+	if err != nil {
+		t.Fatalf("error creating request: %v", err)
+	}
+
+	client := server.Client()
+	client.Timeout = 30 * time.Second
+	_, err = client.Do(req)
+	if err != nil {
+		t.Fatalf("Unexpected non-nil err from client.Do: %v", err)
+	}
+}
+
+func assertCloseNotifierFlusherHijacker(t *testing.T, hijackableExpected bool, w http.ResponseWriter) {
+	// the http.ResponseWriter object for both http/1.x and http2
+	// implement http.Flusher and http.CloseNotifier
+	if _, ok := w.(http.Flusher); !ok {
+		t.Errorf("Expected the http.ResponseWriter object to implement http.Flusher")
+	}
+
+	//lint:ignore SA1019 backward compatibility
+	if _, ok := w.(http.CloseNotifier); !ok {
+		t.Errorf("Expected the http.ResponseWriter object to implement http.CloseNotifier")
+	}
+
+	// http/1.x implements http.Hijacker, not http2
+	if _, ok := w.(http.Hijacker); ok != hijackableExpected {
+		t.Errorf("Unexpected http.Hijacker implementation, expected: %t, but got: %t", hijackableExpected, ok)
+	}
+}
+
+type counter struct {
+	FlushInvoked       int
+	HijackInvoked      int
+	CloseNotifyInvoked int
+}
+
+func (c *counter) count(call string) {
+	switch call {
+	case "Flush":
+		c.FlushInvoked++
+	case "Hijack":
+		c.HijackInvoked++
+	case "CloseNotify":
+		c.CloseNotifyInvoked++
+	}
+}
+
+func (c *counter) assert(t *testing.T, flushInvokedExpected, closeNotifyInvokedExpected, hijackInvokedExpected int) {
+	if flushInvokedExpected != c.FlushInvoked {
+		t.Errorf("Expected Flush() count to match, expected: %d, but got: %d", flushInvokedExpected, c.FlushInvoked)
+	}
+	if closeNotifyInvokedExpected != c.CloseNotifyInvoked {
+		t.Errorf("Expected CloseNotify() count to match, expected: %d, but got: %d", closeNotifyInvokedExpected, c.CloseNotifyInvoked)
+	}
+	if hijackInvokedExpected != c.HijackInvoked {
+		t.Errorf("Expected Hijack() count to match, expected: %d, but got: %d", hijackInvokedExpected, c.HijackInvoked)
+	}
+}
+
+type fakeResponseWriterDecorator struct {
+	http.ResponseWriter
+	callback func(string)
```

Comment:
This fake is only used in this test file.  We do not need the complexity of a general callback here; instead, we could just put the three counters here.

Reply:
if I move the counter inside `fakeResponseWriterDecorator` then I will need to write a test helper function that goes through the chain of decorator and do casting to get to the individual counter and then sum it up. So we are just moving the complexity to a different site.

Reply:
We could just put a pointer to the counters here, and make the CloseNotify, Flush, and Hijack methods of fakeResponseWriterDecorator just directly increment the relevant counter.  That would be better that encoding the desired operation as a string, and then elsewhere decoding the string into the desired increment operation.